File: sdk/typescript/examples/download.ts
/**
 * SafeShare SDK - Download Example
 *
 * Demonstrates file download with progress tracking.
 *
 * Usage:
 *   npx tsx examples/download.ts <claim-code> [destination]
 */

import { SafeShareClient, NotFoundError, PasswordRequiredError } from "../src/index.js";
import * as readline from "node:readline";

async function main() {
  const claimCode = process.argv[2];
  const destination = process.argv[3];

  if (!claimCode) {
    console.error("Usage: npx tsx examples/download.ts <claim-code> [destination]");
    process.exit(1);
  }

  // Create client
  const client = new SafeShareClient({
    baseUrl: process.env.SAFESHARE_URL || "http://localhost:8080",
    apiToken: process.env.SAFESHARE_TOKEN,
  });

  console.log(`Fetching file info for: ${claimCode}\n`);

  try {
    // Get file info first
    const info = await client.getFileInfo(claimCode);

    console.log("File Information:");
    console.log("─".repeat(40));
    console.log(`Filename: ${info.filename}`);
    console.log(`Size: ${formatBytes(info.size)}`);
    console.log(`MIME Type: ${info.mimeType}`);
    console.log(`Expires: ${info.expiresAt || "Never"}`);
    console.log(`Downloads Remaining: ${info.downloadsRemaining ?? "Unlimited"}`);
    console.log(`Password Protected: ${info.passwordProtected}`);
    console.log("─".repeat(40));

    // Determine destination path
    const destPath = destination || `./${info.filename}`;
    console.log(`\nDownloading to: ${destPath}`);

    // Get password if needed
    let password: string | undefined;
    if (info.passwordProtected) {
      password = await promptPassword("Enter password: ");
    }

    // Download with progress
    const savedPath = await client.download(claimCode, destPath, {
      password,
      onProgress: (progress) => {
        const bar = createProgressBar(progress.percentage);
        const current = formatBytes(progress.bytesDownloaded);
        const total = progress.totalBytes > 0 ? formatBytes(progress.totalBytes) : "?";
        const pct = progress.percentage >= 0 ? `${progress.percentage}%` : "?%";

        process.stdout.write(`\r${bar} ${pct} (${current}/${total})`);
      },
    });

    console.log("\n\nDownload complete!");
    console.log(`Saved to: ${savedPath}`);
  } catch (error) {
    if (error instanceof NotFoundError) {
      console.error("\nFile not found. It may have expired or been deleted.");
    } else if (error instanceof PasswordRequiredError) {
      console.error("\nIncorrect password.");
    } else {
      console.error("\nDownload failed:", error);
    }
    process.exit(1);
  }
}

function formatBytes(bytes: number): string {
  if (bytes === 0) return "0 B";
  const k = 1024;
  const sizes = ["B", "KB", "MB", "GB", "TB"];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return `${parseFloat((bytes / Math.pow(k, i)).toFixed(2))} ${sizes[i]}`;
}

function createProgressBar(percentage: number, width = 30): string {
  if (percentage < 0) {
    return `[${"?".repeat(width)}]`;
  }
  const filled = Math.round((percentage / 100) * width);
  const empty = width - filled;
  return `[${"█".repeat(filled)}${"░".repeat(empty)}]`;
}

function promptPassword(prompt: string): Promise<string> {
  return new Promise((resolve) => {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });

    rl.question(prompt, (answer) => {
      rl.close();
      resolve(answer);
    });
  });
}

main();

File: sdk/go/cmd/safeshare-cli/download.go
package main

import (
	"context"
	"fmt"
	"strings"

	safeshare "github.com/fjmerc/safeshare/sdk/go"
	"github.com/spf13/cobra"
)

func downloadCmd() *cobra.Command {
	var (
		password   string
		noProgress bool
	)

	cmd := &cobra.Command{
		Use:   "download <claim-code> [destination]",
		Short: "Download a file by claim code",
		Long: `Download a file from SafeShare using its claim code.

If no destination is specified, the file is saved to the current directory
with its original filename.

Examples:
  safeshare-cli download abc12345
  safeshare-cli download abc12345 ./myfile.pdf
  safeshare-cli download abc12345 --password secret`,
		Args: cobra.RangeArgs(1, 2),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := checkConfig(); err != nil {
				return err
			}

			claimCode := args[0]

			// Create client
			client, err := safeshare.NewClient(safeshare.ClientConfig{
				BaseURL:  baseURL,
				APIToken: apiToken,
			})
			if err != nil {
				return err
			}

			ctx := context.Background()

			// Get file info first
			info, err := client.GetFileInfo(ctx, claimCode)
			if err != nil {
				return fmt.Errorf("getting file info: %w", err)
			}

			// Determine destination
			destination := info.Filename
			if len(args) > 1 {
				destination = args[1]
			}

			fmt.Println("File Information:")
			fmt.Println(strings.Repeat("─", 40))
			fmt.Printf("Filename:    %s\n", info.Filename)
			fmt.Printf("Size:        %s\n", formatBytes(info.Size))
			fmt.Printf("MIME Type:   %s\n", info.MimeType)
			if info.ExpiresAt != nil {
				fmt.Printf("Expires:     %s\n", info.ExpiresAt.Format("2006-01-02 15:04:05"))
			} else {
				fmt.Printf("Expires:     Never\n")
			}
			if info.DownloadsRemaining != nil {
				fmt.Printf("Downloads:   %d remaining\n", *info.DownloadsRemaining)
			} else {
				fmt.Printf("Downloads:   Unlimited\n")
			}
			fmt.Printf("Password:    %v\n", info.PasswordProtected)
			fmt.Println(strings.Repeat("─", 40))
			fmt.Printf("\nDownloading to: %s\n", destination)

			// Prepare options
			opts := &safeshare.DownloadOptions{}
			if password != "" {
				opts.Password = password
			}

			// Progress callback
			if !noProgress {
				opts.OnProgress = func(p safeshare.DownloadProgress) {
					bar := progressBar(p.Percentage)
					if p.Percentage >= 0 {
						fmt.Printf("\r%s %3d%% (%s/%s)",
							bar,
							p.Percentage,
							formatBytes(p.BytesDownloaded),
							formatBytes(p.TotalBytes),
						)
					} else {
						fmt.Printf("\rDownloading... %s", formatBytes(p.BytesDownloaded))
					}
				}
			}

			// Download
			if err := client.Download(ctx, claimCode, destination, opts); err != nil {
				fmt.Println() // Clear progress line
				return err
			}

			fmt.Println() // Clear progress line
			fmt.Println("\nDownload complete!")

			return nil
		},
	}

	cmd.Flags().StringVarP(&password, "password", "p", "", "Password for protected files")
	cmd.Flags().BoolVar(&noProgress, "no-progress", false, "Disable progress bar")

	return cmd
}

func infoCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "info <claim-code>",
		Short: "Get information about a file",
		Long: `Get information about a file without downloading it.

Example:
  safeshare-cli info abc12345`,
		Args: cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := checkConfig(); err != nil {
				return err
			}

			claimCode := args[0]

			// Create client
			client, err := safeshare.NewClient(safeshare.ClientConfig{
				BaseURL:  baseURL,
				APIToken: apiToken,
			})
			if err != nil {
				return err
			}

			ctx := context.Background()
			info, err := client.GetFileInfo(ctx, claimCode)
			if err != nil {
				return err
			}

			fmt.Println("File Information:")
			fmt.Println(strings.Repeat("─", 40))
			fmt.Printf("Filename:    %s\n", info.Filename)
			fmt.Printf("Size:        %s\n", formatBytes(info.Size))
			fmt.Printf("MIME Type:   %s\n", info.MimeType)
			if info.ExpiresAt != nil {
				fmt.Printf("Expires:     %s\n", info.ExpiresAt.Format("2006-01-02 15:04:05"))
			} else {
				fmt.Printf("Expires:     Never\n")
			}
			if info.DownloadsRemaining != nil {
				fmt.Printf("Downloads:   %d remaining\n", *info.DownloadsRemaining)
			} else {
				fmt.Printf("Downloads:   Unlimited\n")
			}
			fmt.Printf("Password:    %v\n", info.PasswordProtected)
			fmt.Println(strings.Repeat("─", 40))

			return nil
		},
	}

	return cmd
}
